### PR TITLE
PEP 680: "tomllib" Support for parsing TOML in the Standard Library

### PR DESCRIPTION
**Pull request by hauntsaninja**

### CODE REVIEW

File: pep-9999.rst [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+Including an API for writing TOML
+---------------------------------
+
+There are several reasons to not include an API for writing TOML:
+
+The ability to write TOML is not needed for the use cases that motivate this
+PEP: for core Python packaging use cases or for tools that need to read
+configuration.
+
+Use cases that involve editing TOML (as opposed to writing brand new TOML) are
```
Comment (sobolevn) [inline pep-9999.rst:231]: Do you refer for cases like `poetry add` here? For example, `poetry add requests` will write `requests = "^x.y"` to your `pyproject.toml` file.

Comment (hukkin) [reply pep-9999.rst:231]: Yes, that's a good example.

**Comment by encukou** [comment]
The next available number is PEP 680. Could you use that, to make the PEP editors' work easier?
Preferably check the [PRs](https://github.com/python/peps/pulls) beforehand: if someone takes 680, get the next number.

**Review by encukou — APPROVED** [review]
I'm happy to sponsor this. Some of the details aren't what I'd choose, but it's a good draft PEP.

File: pep-9999.rst [hunk 2]
```diff
@@ -0,0 +1,501 @@
+PEP: 9999
+Title: tomllib: Support for parsing TOML in the Standard Library
+Author: Taneli Hukkinen, Shantanu Jain <hauntsaninja at gmail.com>
```
Comment (hukkin) [inline pep-9999.rst:3]: The first author should be Shantanu Jain, assuming it reflects workload on this PEP document. If on the other hand we also take into account things like authoring the library and integrating it to the stdlib, then I'm happy with the current order.

Comment (JelleZijlstra) [reply pep-9999.rst:3]: That's something for the two of you as the PEP authors to decide, I don't think there's any strong norm.

File: pep-9999.rst [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+library helps avoid bootstrapping problems for Python build tools. Currently
+most Python build tools need to vendor a TOML parsing library.
+
+Python tools are increasingly configurable via TOML, for examples: ``black``,
+``mypy``, ``pytest``, ``tox``, ``pylint``, ``isort``. Those that are not, such
+as ``flake8``, cite the lack of standard library support as a `main reason why
+<https://github.com/PyCQA/flake8/issues/234#issuecomment-812800657>`_.
+
+Given the special place TOML already has in the Python ecosystem, it makes sense
+for this to be an included battery.
```
Comment (warsaw) [inline pep-9999.rst:36]: I would suggest getting buy-in from the maintainers of tools that support `pyproject.toml`, e.g. poetry, pdm, and the tools mentioned above.

Comment (hukkin) [reply pep-9999.rst:36]: `mypy` should be easy seeing that @hauntsaninja is a maintainer :smile: 

`flit` is not currently listed but I'm fairly sure I won't have to persuade @takluyver ?

`isort` has a no-dependencies policy (vendor instead) so I bet @timothycrosley also has no objections?

Comment (JelleZijlstra) [reply pep-9999.rst:36]: As a maintainer of Black and mypy I'd also be happy to see a TOML parser in the stdlib :)

File: pep-0680.rst [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+standard library backport, we would still break users who have pinned to
+versions of the current ``toml`` package. This is unfortunate, since pinning
+would likely be a common response to breaking changes introduced by repurposing
+the ``toml`` package as a backport (that is incompatible with today's ``toml``).
+
+There are several API incompatibilities between ``toml`` and the API proposed in
+this PEP, listed in `Appendix A`_.
+
+Finally, the ``toml`` package on PyPI is not actively maintained and `we have
+been unable to contact the author <https://github.com/uiri/toml/issues/361>`,
```
Comment (merwok) [inline pep-0680.rst:342]: underscores are missing to make this a link

File: pep-0680.rst [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+#. ``toml.load[s]`` support an undocumented ``decoder`` argument
+
+   It seems the intended use case is for an implementation of comment
+   preservation. The information recorded is not sufficient to roundtrip the
+   TOML document preserving style, the implementation has known bugs, the
+   feature is undocumented and I could only find one instance of its use on
+   https://grep.app.
+
+   The ``toml.TomlDecoder`` interface exposed is not simple, containing nine methods.
+   See `here <https://github.com/uiri/toml/blob/3f637dba5f68db63d4b30967fedda51c82459471/toml/decoder.pyi#L36>`__.
```
Comment (merwok) [inline pep-0680.rst:450]: Using `here` or `click` for text links is poor metadata and accessibility

File: pep-9999.rst (hunk 3 above)
Comment (hugovk) [reply pep-9999.rst]: I expect @pradyunsg could do so for pip and some other packaging tools?

pip switched vendoring toml to tomli: https://github.com/pypa/pip/pull/10035

Comment (pfmoore) [reply pep-9999.rst]: I'd be in favour as a pip maintainer.

File: pep-0680.rst (hunk 5 above)
Comment (hukkin) [reply pep-0680.rst:450]: Fixed [here](https://github.com/python/peps/pull/2223)

(haha :smile: )

File: pep-0680.rst (hunk 4 above)
Comment (hukkin) [reply pep-0680.rst:342]: Addressed this here:  `https://github.com/python/peps/pull/2223`

**Pull request merged**